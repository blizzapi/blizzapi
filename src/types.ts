/** Common */

export type Uri = string;

/** Generic */

export enum ConstantKey {
  REGIONS = "REGIONS",
  LOCALES = "LOCALES",
  DEFAULT_LOCALES = "DEFAULT_LOCALES",
  REGION_API_HOSTS = "REGION_API_HOSTS",
  SC2_REALMS = "SC2_REALMS",
  OAUTH_AUTHORIZE_URIS = "OAUTH_AUTHORIZE_URIS",
  OAUTH_TOKEN_URIS = "OAUTH_TOKEN_URIS",
  OAUTH_CHECK_TOKEN_URIS = "OAUTH_CHECK_TOKEN_URIS",
}

/** Regions */

export enum RegionName {
  us = "us",
  eu = "eu",
  kr = "kr",
  tw = "tw",
  cn = "cn",
}

export enum RegionId {
  us = 1,
  eu = 2,
  kr = 3,
  tw = 3,
  cn = 5,
}

export enum RegionIdAsString {
  us = "1",
  eu = "2",
  kr = "3",
  tw = "3",
  cn = "5",
}

export type RegionIdKey = string;
export type RegionIdAsNumberOrString = RegionId | RegionIdAsString;
export type RegionIdArray = ReadonlyArray<RegionId>;
export type RegionNameArray = ReadonlyArray<RegionName>;

/** Locales */

export enum Locale {
  en_US = "en_US",
  es_MX = "es_MX",
  pt_BR = "pt_BR",
  en_GB = "en_GB",
  es_ES = "es_ES",
  fr_FR = "fr_FR",
  ru_RU = "ru_RU",
  de_DE = "de_DE",
  pt_PT = "pt_PT",
  it_IT = "it_IT",
  ko_KR = "ko_KR",
  zh_TW = "zh_TW",
  zh_CN = "zh_CN",
}

export type LocaleArray = ReadonlyArray<Locale>;

/** Default locales */

export enum DefaultLocaleIndex {
  us = 0,
  eu = 0,
  kr = 0,
  tw = 0,
  cn = 0,
}

export type DefaultLocaleIndexArray = ReadonlyArray<DefaultLocaleIndex>;

/** StarCraft II realms  */

export enum Sc2Realm {
  US = 1,
  LatAm = 2,
  Europe = 1,
  Russia = 2,
  Korea = 1,
  Taiwan = 2,
  China = 1,
}

export enum Sc2RealmAsString {
  US = "1",
  LatAm = "2",
  Europe = "1",
  Russia = "2",
  Korea = "1",
  Taiwan = "2",
  China = "1",
}

export type Sc2RealmArray = ReadonlyArray<Sc2Realm>;
export type Sc2RealmList = ReadonlyArray<Sc2Realm>;
export type Sc2RealmAsNumberOrString = Sc2Realm | Sc2RealmAsString;

/** API class types */

export type RegionIdOrName = RegionId | RegionName;
export type MaybeRegion = RegionIdOrName | undefined;
export type ClientId = string;
export type ClientSecret = string;
export type AccessToken = string | undefined;
export type Endpoint = string;

export enum HttpMethod {
  GET = "GET",
  POST = "POST",
}

export type ValidatorFunction = (endpoint: Endpoint) => boolean;

/* OAuth flow URIs */

export enum OAuthHost {
  us = "https://us.battle.net",
  eu = "https://eu.battle.net",
  kr = "https://apac.battle.net",
  tw = "https://apac.battle.net",
  cn = "https://www.battlenet.com.cn",
}

export enum RegionHost {
  us = "https://us.api.blizzard.com",
  eu = "https://eu.api.blizzard.com",
  kr = "https://kr.api.blizzard.com",
  tw = "https://tw.api.blizzard.com",
  cn = "https://gateway.battlenet.com.cn",
}

export enum OAuthEndpoint {
  authorize = "/oauth/authorize",
  token = "/oauth/token",
  checkToken = "/oauth/check_token?token=",
}

/** API headers */

export enum ApiHeaders {
  LastModified = "last-modified",
}

/** Errors */

export enum ErrorCode {
  NotAuthorized = 401,
}

export enum ErrorResponseMessage {
  InvalidToken = "invalid_token",
  AccessTokenInvalid = "access_token_invalid",
  AccessTokenExpired = "access_token_invalid",
}

export type InvalidAccessTokenError = {
  error: ErrorResponseMessage.AccessTokenInvalid;
};

export type ExpiredAccessTokenError = {
  error: ErrorResponseMessage.AccessTokenExpired;
};

export type ResponseError = InvalidAccessTokenError | ExpiredAccessTokenError;

/** Interfaces */

export interface AccessTokenOptions {
  validateAccessTokenOnEachQuery?: boolean;
  refreshExpiredAccessToken?: boolean;
  onAccessTokenExpired?: () => void;
  onAccessTokenRefresh?: (newAccessToken: string) => void;
}

export interface QueryOptions {
  timeout?: number;
  headers?: object | Headers;
  params?: object | URLSearchParams;
}

export interface BattleNetOptions {
  region: RegionIdOrName;
  clientId: ClientId;
  clientSecret: ClientSecret;
  accessToken?: AccessToken;
}

export interface InitOptions {
  region: string | number;
  clientId?: string;
  clientSecret?: string;
  accessToken?: string;
  validateAccessTokenOnEachQuery?: boolean;
  revalidateAccessTokenIfExpired?: boolean;
  onAccessTokenExpire?: () => void;
  onAccessTokenRevalidate?: () => void;
}

export interface BattleNetQueryOptions {
  region: RegionIdOrName;
  endpoint: string;
  clientId: string;
  clientSecret: string;
  accessToken: AccessToken;
  options: AccessTokenOptions & QueryOptions;
}

export type RegionIdProperties<Value> = {
  [key in RegionId]: Value;
};

export interface BlizzAPIOptions extends BattleNetOptions, AccessTokenOptions {}
