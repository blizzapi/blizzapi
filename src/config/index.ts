/**
 * Class exposing configuration constants
 *
 * Via:
 *
 * https://develop.battle.net/documentation/guides/using-oauth
 * https://develop.battle.net/documentation/guides/regionality-partitions-and-localization
 */

import * as localization from 'config/localization';
import * as api from 'config/api';
import oauth from 'config/oauth';
import { ConfigObject } from 'types';

const config = {
  REGIONS: localization.regions,
  LOCALES: localization.locales,
  DEFAULT_LOCALES: localization.defaultLocales,
  SC2_REALMS: localization.sc2Realms,
  REGION_API_HOSTS: api.regionHosts,
  OAUTH_AUTHORIZE_URIS: oauth.authorizeUris,
  OAUTH_TOKEN_URIS: oauth.tokenUris,
  OAUTH_CHECK_TOKEN_URIS: oauth.checkTokenUris,
} as ConfigObject;

export default config;
