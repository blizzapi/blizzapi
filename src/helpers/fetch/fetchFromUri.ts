import axios, { AxiosBasicCredentials, AxiosRequestConfig } from "axios";
import { uri as validateUri } from "../validators";
import { Uri, HttpMethod, QueryOptions } from "../../types";

interface FetchFromUriOptions extends QueryOptions {
  uri: Uri;
  method?: HttpMethod;
  data?: string;
  auth?: AxiosBasicCredentials;
}

/**
 * Performs basic fetch request from a given uri
 */
export const fetchFromUri = async (options: FetchFromUriOptions) => {
  const { uri, timeout, headers, params, data, auth } = options;

  const method = options.method || "GET";

  if (!validateUri(uri)) {
    throw new RangeError(
      `'${uri}' is not a valid parameter for fetchFromUri()`
    );
  }

  const requestOptions = {
    method,
    url: encodeURI(uri),
    timeout: timeout || 10000,
    ...(headers && { headers }),
    ...(params && { params }),
    ...(auth && { auth }),
    ...(data && { data }),
  };

  const response = await axios.request(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    requestOptions as AxiosRequestConfig<any>
  );

  const lastModified = response.headers["last-modified"]
    ? response.headers["last-modified"]
    : null;

  return {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    ...(response.data as any),
    ...(lastModified && { lastModified }),
  };
};
