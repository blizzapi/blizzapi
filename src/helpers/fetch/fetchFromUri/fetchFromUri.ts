import axios from 'axios';
import { HttpMethod } from '../../../../types';
import { uri as validateUri } from '../../validators';

interface FetchFromUriOptions {
  uri: string,
  method?: HttpMethod;
  headers?: object | Headers;
  params?: object;
}

/**
 * Performs basic fetch request from a given uri
 */
export default async (options: FetchFromUriOptions) => {
  const {
    uri,
    method,
    headers, 
    params,
  } = options;

  const encodedUri = encodeURI(uri);
  const fetchMethod = method || 'GET';
  try {
    if (!validateUri(encodedUri)) {
      throw new RangeError(`'${uri}' is not a valid parameter for fetchFromUri()`);
    }
    const requestOptions = {
      method,
    };

    /* tslint:disable no-object-mutation */
    if (headers) Object.assign(requestOptions, { headers });

    // GET request method cannot have body, so I'm doing this nonsense
    if (fetchMethod === 'POST') Object.assign(requestOptions, { params });
    /* tslint:enable no-object-mutation */
    const response = await axios.get(uri, requestOptions);
    return response.data;
  } catch (error) {
    throw error;
  }
};
