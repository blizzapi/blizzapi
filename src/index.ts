import BlizzAPI from './lib/classes/BlizzAPI';
export = BlizzAPI;