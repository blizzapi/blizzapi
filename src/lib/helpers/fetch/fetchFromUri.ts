import axios from 'axios';
import { Uri, HttpMethod } from '../../../../@types';
import { uri as validateUri } from '../validators';

interface FetchFromUriOptions {
  uri: Uri,
  method?: HttpMethod;
  headers?: object | Headers;
  params?: object;
}

/**
 * Performs basic fetch request from a given uri
 */
export default async (options: FetchFromUriOptions) => {
  const uri = encodeURI(options.uri);
  const method = options.method || 'GET';
  try {
    if (!validateUri(uri)) {
      throw new RangeError(`'${uri}' is not a valid parameter for fetchFromUri()`);
    }
    const requestOptions = {
      method,
    };

    /* tslint:disable no-object-mutation */
    if (options.headers) Object.assign(requestOptions, { headers: options.headers });

    // GET request method cannot have body, so I'm doing this nonsense
    if (method === 'POST') Object.assign(requestOptions, { params: options.params });
    /* tslint:enable no-object-mutation */
    const response = await axios.get(uri, requestOptions);
    return response.data;
  } catch (error) {
    throw error;
  }
};
