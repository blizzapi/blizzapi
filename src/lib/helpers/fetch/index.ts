export { default as fetchFromUri } from './fetchFromUri';
