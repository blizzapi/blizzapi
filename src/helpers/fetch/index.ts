import fetchFromUri from './fetchFromUri/fetchFromUri';
export {
  fetchFromUri,
}
