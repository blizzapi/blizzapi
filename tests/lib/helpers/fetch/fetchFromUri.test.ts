import { fetchFromUri } from "../../../../src/helpers/fetch/fetchFromUri";
import { HttpMethod } from "../../../../src/types";

describe("fetchFromUri()", () => {
  it("should be defined", () => {
    expect(fetchFromUri).toBeDefined();
  });

  it("should be function", () => {
    expect(fetchFromUri).toBeTruthy();
    expect(typeof fetchFromUri).toBe("function");
  });

  it("returns correct response for GET request to a valid uri", async () => {
    expect.assertions(1);
    const response = await fetchFromUri({
      uri: "http://example.org/valid/endpoint",
    });
    expect(response).toMatchSnapshot();
  });

  it("returns correct response for POST request to a valid uri", async () => {
    expect.assertions(1);
    const response = await fetchFromUri({
      uri: "http://example.org/valid/endpoint",
      method: "POST" as HttpMethod,
    });
    expect(response).toMatchSnapshot();
  });

  it("returns correct response for GET request with headers to a valid uri", async () => {
    expect.assertions(1);
    const response = await fetchFromUri({
      uri: "http://example.org/valid/endpoint",
      method: "GET" as HttpMethod,
      headers: {
        data: "sample headers",
      },
    });
    expect(response).toMatchSnapshot();
  });

  it("returns correct response for GET request to uri containing regional characters", async () => {
    expect.assertions(1);
    const response = await fetchFromUri({
      uri: "http://ęxąmple.org/zażółćgęśląjaźń/гордунни/좋은경기",
      method: "GET" as HttpMethod,
    });
    expect(response).toMatchSnapshot();
  });

  it("returns correct response for GET request with params to a valid uri", async () => {
    expect.assertions(1);
    const response = await fetchFromUri({
      uri: "http://example.org/valid/endpoint",
      method: "GET" as HttpMethod,
      params: {
        data: "custom params",
      },
    });
    expect(response).toMatchSnapshot();
  });

  it('returns correct response for GET request with "last-modified" header', async () => {
    expect.assertions(1);
    const response = await fetchFromUri({
      uri: "http://example.org/valid/endpoint/lastModified",
      method: "GET" as HttpMethod,
      params: {
        data: "custom params",
      },
    });
    expect(response).toMatchSnapshot();
  });
});
