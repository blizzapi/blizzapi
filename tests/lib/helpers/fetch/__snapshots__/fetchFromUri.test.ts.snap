// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchFromUri() returns correct response for GET request to a valid uri 1`] = `
Object {
  "access_token": "no_access_token_provided",
  "headers": Object {
    "data": "default sample headers",
  },
  "params": Object {
    "data": "default sample params",
  },
  "title": "sample mocked data for GET request to http://example.org/valid/endpoint",
}
`;

exports[`fetchFromUri() returns correct response for GET request to uri containing regional characters 1`] = `
Object {
  "access_token": "no_access_token_provided",
  "headers": Object {
    "data": "default sample headers",
  },
  "params": Object {
    "data": "default sample params",
  },
  "title": "sample mocked data for GET request to http://%C4%99x%C4%85mple.org/za%C5%BC%C3%B3%C5%82%C4%87g%C4%99%C5%9Bl%C4%85ja%C5%BA%C5%84/%D0%B3%D0%BE%D1%80%D0%B4%D1%83%D0%BD%D0%BD%D0%B8/%EC%A2%8B%EC%9D%80%EA%B2%BD%EA%B8%B0",
}
`;

exports[`fetchFromUri() returns correct response for GET request with headers to a valid uri  1`] = `
Object {
  "access_token": "no_access_token_provided",
  "headers": Object {
    "data": "sample headers",
  },
  "params": Object {
    "data": "default sample params",
  },
  "title": "sample mocked data for GET request to http://example.org/valid/endpoint",
}
`;

exports[`fetchFromUri() returns correct response for GET request with params to a valid uri  1`] = `
Object {
  "access_token": "no_access_token_provided",
  "headers": Object {
    "data": "default sample headers",
  },
  "params": Object {
    "data": "custom params",
  },
  "title": "sample mocked data for GET request to http://example.org/valid/endpoint",
}
`;

exports[`fetchFromUri() returns correct response for POST request to a valid uri 1`] = `
Object {
  "access_token": "no_access_token_provided",
  "headers": Object {
    "data": "default sample headers",
  },
  "params": Object {
    "data": "default sample params",
  },
  "title": "sample mocked data for POST request to http://example.org/valid/endpoint",
}
`;
